File: packages/catsproto/test/grid_test.dart
import 'package:catsproto/catsproto.dart';
import 'package:test/test.dart';

PaneFrame frame({
  int w = 4,
  int h = 3,
  int defFg = 0x02c0c0c0,
  int defBg = 0x02000000,
}) => PaneFrame(
  pane: 1,
  w: w,
  h: h,
  cur: const Cursor(x: 0, y: 0, vis: true, shape: 0),
  defFg: defFg,
  defBg: defBg,
  cells: List<Cell>.generate(
    w * h,
    (i) => Cell(s: String.fromCharCode(97 + i % 26)),
  ),
  scroll: const Scroll(off: 0, max: 100, rows: 3),
);

void main() {
  group('PaneGrid.applyFrame', () {
    test('resolves omitted colours against the frame defaults', () {
      final grid = PaneGrid(1)..applyFrame(frame());
      // Every cell in the fixture omits both colours, which is the dominant
      // case on the wire — resolving here means the painter reads a concrete
      // value instead of branching 3840 times a frame.
      expect(grid.fg.every((c) => c == 0x02c0c0c0), isTrue);
      expect(grid.bg.every((c) => c == 0x02000000), isTrue);
    });

    test('keeps an explicit colour', () {
      final f = frame();
      final grid = PaneGrid(1)
        ..applyFrame(
          PaneFrame(
            pane: 1,
            w: f.w,
            h: f.h,
            cur: f.cur,
            defFg: f.defFg,
            defBg: f.defBg,
            cells: [
              const Cell(s: 'x', f: 0x02ff0000),
              ...f.cells.skip(1),
            ],
          ),
        );
      expect(grid.fg[0], 0x02ff0000);
      expect(grid.fg[1], f.defFg);
    });

    test('marks every row dirty', () {
      final grid = PaneGrid(1)..applyFrame(frame());
      expect(grid.dirtyRows.every((r) => r == 1), isTrue);
      grid.clearDirty();
      expect(grid.dirtyRows.every((r) => r == 0), isTrue);
    });

    test(
      'a short cells array blanks the tail instead of showing stale content',
      () {
        final grid = PaneGrid(1)..applyFrame(frame());
        grid.applyFrame(
          PaneFrame(
            pane: 1,
            w: 4,
            h: 3,
            cur: const Cursor(x: 0, y: 0, vis: false, shape: 0),
            defFg: 1,
            defBg: 2,
            cells: const [Cell(s: 'z')],
          ),
        );
        expect(grid.glyph[0], 'z');
        expect(grid.glyph.skip(1).every((g) => g.isEmpty), isTrue);
      },
    );
  });

  group('PaneGrid.applyDiff', () {
    test('a cell at index i dirties row i ~/ w and no other', () {
      final grid = PaneGrid(1)..applyFrame(frame(w: 4, h: 3));
      grid.clearDirty();
      grid.applyDiff(const PaneDiff(pane: 1, cells: [DiffCell(i: 6, s: 'Z')]));
      expect(grid.rowOf(6), 1);
      expect(grid.dirtyRows[0], 0);
      expect(grid.dirtyRows[1], 1);
      expect(grid.dirtyRows[2], 0);
      expect(grid.glyph[6], 'Z');
    });

    test('a diff before any full frame is dropped AND counted', () {
      // Counted rather than silently ignored: registerConn guarantees a full
      // frame per visible pane on connect, so a non-zero count means something
      // is wrong with the resync, not with the wire.
      final grid = PaneGrid(1);
      grid.applyDiff(
        const PaneDiff(
          pane: 1,
          cells: [
            DiffCell(i: 0, s: 'a'),
            DiffCell(i: 1, s: 'b'),
          ],
        ),
      );
      expect(grid.hasFullFrame, isFalse);
      expect(grid.droppedCells, 2);
    });

    test('an out-of-range index is dropped and counted', () {
      final grid = PaneGrid(1)..applyFrame(frame(w: 4, h: 3));
      grid.applyDiff(
        const PaneDiff(
          pane: 1,
          cells: [
            DiffCell(i: 999, s: 'x'),
            DiffCell(i: -1, s: 'y'),
          ],
        ),
      );
      expect(grid.droppedCells, 2);
    });

    test('a null scroll CLEARS rather than preserving the last one', () {
      // A stale `max` misplaces the absolute coordinates a later `read` is
      // computed from, which selects the wrong text without any error.
      final grid = PaneGrid(1)..applyFrame(frame());
      expect(grid.scroll?.max, 100);
      grid.applyDiff(const PaneDiff(pane: 1, cells: []));
      expect(grid.scroll, isNull);
    });

    test(
      'omitted diff colours resolve against the LAST FULL frame defaults',
      () {
        final grid = PaneGrid(1)
          ..applyFrame(frame(defFg: 0x02111111, defBg: 0x02222222));
        grid.applyDiff(
          const PaneDiff(pane: 1, cells: [DiffCell(i: 2, s: 'q')]),
        );
        expect(grid.fg[2], 0x02111111);
        expect(grid.bg[2], 0x02222222);
      },
    );

    test('a cursor is only replaced when the diff carries one', () {
      final grid = PaneGrid(1)..applyFrame(frame());
      grid.applyDiff(const PaneDiff(pane: 1, cells: []));
      expect(grid.cursor.vis, isTrue, reason: 'an absent cur means unchanged');
      grid.applyDiff(
        const PaneDiff(
          pane: 1,
          cur: Cursor(x: 3, y: 2, vis: false, shape: 1),
          cells: [],
        ),
      );
      expect(grid.cursor.x, 3);
      expect(grid.cursor.vis, isFalse);
    });
  });

  group('wide glyphs', () {
    test('the tail cell is empty, not a space, and contributes no text', () {
      // Verified against the server with:
      //   catctl probe --script 'type:printf "AB你好CD|\n"; dump:1'
      // The wire grid is always exactly w*h; a double-width glyph occupies its
      // own cell and leaves the next one as "" — which is what lets the glyph
      // spill into it at paint time instead of being clipped.
      final grid = PaneGrid(1)
        ..applyFrame(
          PaneFrame(
            pane: 1,
            w: 4,
            h: 1,
            cur: const Cursor(x: 0, y: 0, vis: false, shape: 0),
            defFg: 1,
            defBg: 2,
            cells: const [
              Cell(s: 'A'),
              Cell(s: '你'),
              Cell(s: ''),
              Cell(s: 'B'),
            ],
          ),
        );
      expect(grid.glyph[2], '');
      expect(grid.glyph[2], isNot(' '));
      expect(grid.rowText(0), 'A你B');
    });
  });
}

File: internal/catsclient/grid_test.go
package catsclient

import (
	"testing"

	"github.com/rohanthewiz/cats/wire"
)

// Mirrors packages/catsproto/test/grid_test.dart.

func testFrame(w, h int, defFg, defBg uint32) *wire.PaneFrame {
	cells := make([]wire.Cell, w*h)
	for i := range cells {
		cells[i] = wire.Cell{S: string(rune('a' + i%26))}
	}
	return &wire.PaneFrame{
		T: wire.MsgPaneFrame, Pane: 1, W: uint16(w), H: uint16(h),
		Cur:   wire.Cursor{Vis: true},
		DefFg: defFg, DefBg: defBg,
		Cells:  cells,
		Scroll: &wire.Scroll{Off: 0, Max: 100, Rows: 3},
	}
}

func defaultFrame() *wire.PaneFrame { return testFrame(4, 3, 0x02c0c0c0, 0x02000000) }

func all[T comparable](xs []T, want T) bool {
	for _, x := range xs {
		if x != want {
			return false
		}
	}
	return true
}

func TestApplyFrameResolvesOmittedColoursAgainstTheFrameDefaults(t *testing.T) {
	g := NewPaneGrid(1)
	g.ApplyFrame(defaultFrame())
	// Every cell in the fixture omits both colours, which is the dominant
	// case on the wire; resolving here means the renderer reads a concrete
	// value instead of branching 3840 times a frame.
	if !all(g.Fg(), 0x02c0c0c0) {
		t.Errorf("fg = %v", g.Fg())
	}
	if !all(g.Bg(), 0x02000000) {
		t.Errorf("bg = %v", g.Bg())
	}
}

func TestApplyFrameKeepsAnExplicitColour(t *testing.T) {
	f := defaultFrame()
	f.Cells[0] = wire.Cell{S: "x", F: 0x02ff0000}
	g := NewPaneGrid(1)
	g.ApplyFrame(f)
	if g.Fg()[0] != 0x02ff0000 {
		t.Errorf("fg[0] = %#x", g.Fg()[0])
	}
	if g.Fg()[1] != f.DefFg {
		t.Errorf("fg[1] = %#x, want the default", g.Fg()[1])
	}
}

func TestApplyFrameMarksEveryRowDirty(t *testing.T) {
	g := NewPaneGrid(1)
	g.ApplyFrame(defaultFrame())
	if !all(g.DirtyRows(), true) {
		t.Errorf("dirty = %v", g.DirtyRows())
	}
	g.ClearDirty()
	if !all(g.DirtyRows(), false) {
		t.Errorf("dirty after clear = %v", g.DirtyRows())
	}
}

func TestApplyFrameAShortCellsArrayBlanksTheTail(t *testing.T) {
	// Instead of showing stale content.
	g := NewPaneGrid(1)
	g.ApplyFrame(defaultFrame())
	g.ApplyFrame(&wire.PaneFrame{
		Pane: 1, W: 4, H: 3, DefFg: 1, DefBg: 2,
		Cells: []wire.Cell{{S: "z"}},
	})
	if g.Glyph()[0] != "z" {
		t.Errorf("glyph[0] = %q", g.Glyph()[0])
	}
	if !all(g.Glyph()[1:], "") {
		t.Errorf("tail = %q", g.Glyph()[1:])
	}
	if g.Fg()[5] != 1 || g.Bg()[5] != 2 {
		t.Error("the blanked tail should carry the new defaults")
	}
}

func TestApplyFrameAnOverlongCellsArrayIsClamped(t *testing.T) {
	f := testFrame(2, 1, 1, 2)
	f.Cells = append(f.Cells, wire.Cell{S: "extra"}, wire.Cell{S: "extra"})
	g := NewPaneGrid(1)
	g.ApplyFrame(f)
	if len(g.Glyph()) != 2 {
		t.Errorf("grid has %d cells, want w*h = 2", len(g.Glyph()))
	}
}

func TestApplyDiffACellDirtiesItsRowAndNoOther(t *testing.T) {
	g := NewPaneGrid(1)
	g.ApplyFrame(testFrame(4, 3, 1, 2))
	g.ClearDirty()
	g.ApplyDiff(&wire.PaneDiff{Pane: 1, Cells: []wire.DiffCell{{I: 6, Cell: wire.Cell{S: "Z"}}}})
	if g.RowOf(6) != 1 {
		t.Errorf("rowOf(6) = %d", g.RowOf(6))
	}
	if d := g.DirtyRows(); d[0] || !d[1] || d[2] {
		t.Errorf("dirty = %v, want only row 1", d)
	}
	if g.Glyph()[6] != "Z" {
		t.Errorf("glyph[6] = %q", g.Glyph()[6])
	}
}

func TestApplyDiffBeforeAnyFullFrameIsDroppedAndCounted(t *testing.T) {
	// Counted rather than silently ignored: registerConn guarantees a full
	// frame per visible pane on connect, so a non-zero count means something
	// is wrong with the resync, not with the wire.
	g := NewPaneGrid(1)
	g.ApplyDiff(&wire.PaneDiff{Pane: 1, Cells: []wire.DiffCell{
		{I: 0, Cell: wire.Cell{S: "a"}}, {I: 1, Cell: wire.Cell{S: "b"}},
	}})
	if g.HasFullFrame() {
		t.Error("a diff is not a full frame")
	}
	if g.DroppedCells != 2 {
		t.Errorf("dropped = %d", g.DroppedCells)
	}
}

func TestApplyDiffOutOfRangeIndexIsDroppedAndCounted(t *testing.T) {
	g := NewPaneGrid(1)
	g.ApplyFrame(testFrame(4, 3, 1, 2))
	g.ApplyDiff(&wire.PaneDiff{Pane: 1, Cells: []wire.DiffCell{
		{I: 999, Cell: wire.Cell{S: "x"}}, {I: -1, Cell: wire.Cell{S: "y"}},
	}})
	if g.DroppedCells != 2 {
		t.Errorf("dropped = %d", g.DroppedCells)
	}
}

func TestApplyDiffANilScrollClearsRatherThanPreserving(t *testing.T) {
	// A stale max misplaces the absolute coordinates a later read is computed
	// from, which selects the wrong text without any error.
	g := NewPaneGrid(1)
	g.ApplyFrame(defaultFrame())
	if g.Scroll == nil || g.Scroll.Max != 100 {
		t.Fatalf("scroll = %+v", g.Scroll)
	}
	g.ApplyDiff(&wire.PaneDiff{Pane: 1})
	if g.Scroll != nil {
		t.Errorf("scroll = %+v, want nil", g.Scroll)
	}
}

func TestApplyDiffOmittedColoursResolveAgainstTheLastFullFrame(t *testing.T) {
	g := NewPaneGrid(1)
	g.ApplyFrame(testFrame(4, 3, 0x02111111, 0x02222222))
	g.ApplyDiff(&wire.PaneDiff{Pane: 1, Cells: []wire.DiffCell{{I: 2, Cell: wire.Cell{S: "q"}}}})
	if g.Fg()[2] != 0x02111111 || g.Bg()[2] != 0x02222222 {
		t.Errorf("fg,bg = %#x,%#x", g.Fg()[2], g.Bg()[2])
	}
}

func TestApplyDiffACursorIsOnlyReplacedWhenTheDiffCarriesOne(t *testing.T) {
	g := NewPaneGrid(1)
	g.ApplyFrame(defaultFrame())
	g.ApplyDiff(&wire.PaneDiff{Pane: 1})
	if !g.Cursor.Vis {
		t.Error("an absent cur means unchanged")
	}
	g.ApplyDiff(&wire.PaneDiff{Pane: 1, Cur: &wire.Cursor{X: 3, Y: 2, Vis: false, Shape: 1}})
	if g.Cursor.X != 3 || g.Cursor.Vis {
		t.Errorf("cursor = %+v", g.Cursor)
	}
}

func TestWideGlyphTailCellIsEmptyNotASpace(t *testing.T) {
	// Verified against the server with:
	//   catctl probe --script 'type:printf "AB你好CD|\n"; dump:1'
	// The wire grid is always exactly w*h; a double-width glyph occupies its
	// own cell and leaves the next one as "", which is what lets the glyph
	// spill into it at paint time instead of being clipped.
	g := NewPaneGrid(1)
	g.ApplyFrame(&wire.PaneFrame{
		Pane: 1, W: 4, H: 1, DefFg: 1, DefBg: 2,
		Cells: []wire.Cell{{S: "A"}, {S: "你"}, {S: ""}, {S: "B"}},
	})
	if g.Glyph()[2] != "" {
		t.Errorf("tail = %q, want empty", g.Glyph()[2])
	}
	if got := g.RowText(0); got != "A你B" {
		t.Errorf("rowText = %q", got)
	}
	if g.RowText(5) != "" || g.RowText(-1) != "" {
		t.Error("an out-of-range row is empty, not a panic")
	}
}
